### PR TITLE
feat: add users and transactions to dkg rituals

### PR DESCRIPTION
**Pull request by evandrosaturnino**
**Type of PR:**
- [ ] Bugfix
- [X] Feature
- [ ] Documentation
- [ ] Other

**Required reviews:**
- [x] 1
- [ ] 2
- [ ] 3

**What this does:**
Add `Users` and `Transactions` to the DKG Rituals context
The `Coordinator` mapping file has been refactored following the same code structured used in tBTC Subgraph
Add `publicKey` and `transactions` to the Ritual entity
 
**Why it's needed:**
This PR paves the way for the TACo scan implementation, it adds transactions and users to the DGK rituals context, following the same data pattern implemented in tBTC scan.

Additionally, the main mapping file has been refactored in order to follow the same code structure as the existing tBTC subgraph, providing more similarity between the codebase across different Threshold Network applications.

**Notes for reviewers:**
The resulting subgraph can be checked here:

https://api.studio.thegraph.com/proxy/59264/taco-matic/version/latest/graphql?query=query+MyQuery+%7B%0A++rituals%28first%3A+10%2C+orderBy%3A+initTimestamp%29+%7B%0A++++id%0A++++publicKey%0A++++transactions%28first%3A+10%2C+orderBy%3A+timestamp%29+%7B%0A++++++timestamp%0A++++++description%0A++++++id%0A++++%7D%0A++++dkgStatus%0A++%7D%0A%7D

### CODE REVIEW

File: DKGRituals/src/mappingCoordinator.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+export function handleStartAggregationRound(
+  event: StartAggregationRoundEvent,
+): void {
+  const transaction = getOrCreateTransaction(getTransactionIDFromEvent(event));
+  transaction.txHash = event.transaction.hash;
+  transaction.timestamp = event.block.timestamp;
+  transaction.from = event.transaction.from;
+  transaction.to = event.transaction.to;
+  transaction.description = "Posted Transcripts";
+  transaction.save();
```
Comment (derekpierre) [inline DKGRituals/src/mappingCoordinator.ts:78]: Do you need this? This event is emitted from the Coordinator contract when the last node to post their transcript does so -https://github.com/nucypher/nucypher-contracts/blob/main/contracts/contracts/coordination/Coordinator.sol#L340. So the tx associated with this event is presumably the same as the last `TranscriptPosted` event. Would this cause the tx to be duplicated in the list of transactions?

File: DKGRituals/src/mappingCoordinator.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  ritual.transactions = transactions;
+  ritual.save();
+}
+
+export function handleAggregationPosted(event: AggregationPostedEvent): void {
+  const ritual = getOrCreateRitual(event.params.ritualId.toString());
+  const postedAggregations = ritual.postedAggregations;
+  postedAggregations.push(event.params.node);
+  ritual.postedAggregations = postedAggregations;
+  ritual.postedAggregationsAmount = ritual.postedAggregationsAmount + 1;
```
Comment (derekpierre) [inline DKGRituals/src/mappingCoordinator.ts:94]: Can we do `numPostedAggregations` or `totalPostedAggregations` or something like that - open to others. "Amount"  is typically for things that are not countable.

File: DKGRituals/src/mappingCoordinator.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  ritual.save();
+}
+
+export function handleEndRitual(event: EndRitualEvent): void {
+  const transaction = getOrCreateTransaction(getTransactionIDFromEvent(event));
+  transaction.txHash = event.transaction.hash;
+  transaction.timestamp = event.block.timestamp;
+  transaction.from = event.transaction.from;
+  transaction.to = event.transaction.to;
+  transaction.description = "Posted Aggregations";
```
Comment (derekpierre) [inline DKGRituals/src/mappingCoordinator.ts:104]: Same comment for `EndRitual` as `StartAggregationRound`.

Do we need to store the associated transaction? Presumably it's the same tx as the last `AggregationPosted` tx - https://github.com/nucypher/nucypher-contracts/blob/main/contracts/contracts/coordination/Coordinator.sol#L394 and https://github.com/nucypher/nucypher-contracts/blob/main/contracts/contracts/coordination/Coordinator.sol#L407

File: DKGRituals/src/utils/helper.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import { Bytes } from "@graphprotocol/graph-ts";
```
Comment (derekpierre) [inline DKGRituals/src/utils/helper.ts:1]: Feels like what is in here `helper.ts` can be moved to `utils.ts`.

Any reason for having separate files?

File: DKGRituals/schema.graphql [hunk 5]
```diff
@@ -29,3 +32,18 @@ type RitualCounter @entity {
...
+  id: ID!
+  txHash: Bytes!
+  timestamp: BigInt!
+  from: Bytes!
+  to: Bytes
+  description: String
+  rituals: [Ritual!] @derivedFrom(field: "transactions")
+}
+
+type User @entity {
```
Comment (derekpierre) [inline DKGRituals/schema.graphql:46]: Any obection to renaming this to `Initiator` as opposed to `User`?

File: DKGRituals/src/mappingCoordinator.ts (hunk 1 above)
Comment (evandrosaturnino) [reply DKGRituals/src/mappingCoordinator.ts:78]: Transactions entity is needed here precisely to capture the transaction details from the last tx that would trigger this event.
Essentially It is capturing the tx so we can have a chronological order of transactions that trigger the change of state of the ritual along time. Each ritual has a transactions property and only the tx that triggers the change of state are added into the transactions list of the Ritual entity as you can see it here: https://api.studio.thegraph.com/proxy/59264/taco-matic/version/latest/graphql?query=query+MyQuery+%7B%0A++rituals%28first%3A+10%2C+orderBy%3A+initTimestamp%29+%7B%0A++++id%0A++++publicKey%0A++++transactions%28first%3A+10%2C+orderBy%3A+timestamp%29+%7B%0A++++++timestamp%0A++++++description%0A++++++id%0A++++%7D%0A++++dkgStatus%0A++%7D%0A%7D

File: DKGRituals/src/mappingCoordinator.ts (hunk 3 above)
Comment (evandrosaturnino) [reply DKGRituals/src/mappingCoordinator.ts:104]: The tx of `AggregationPosted` is not being stored, as pointed out on above, the transaction details are only being stored in `Transaction` entity specifically on events that updates the Ritual entity status, maybe worth storing all the transactions, and only add to the Ritual transactions property only those that are relevant for changing the Ritual status?

File: DKGRituals/src/utils/helper.ts (hunk 4 above)
Comment (evandrosaturnino) [reply DKGRituals/src/utils/helper.ts:1]: I'm not strongly opiniated about this, just followed the same code structure pattern used in tBTC subgraphs, at first it made sense using helper.ts for the functions that directly manipulate the graph entities, acting as a helper extension to the mapping handlers. While in utils we can find more generic functions mainly used for transforming data types/values.

File: DKGRituals/schema.graphql (hunk 5 above)
Comment (evandrosaturnino) [reply DKGRituals/schema.graphql]: None, addressed in [update user entity name to initiator](https://github.com/nucypher/taco-subgraphs/pull/10/commits/5daa8b77f8f3e3ccecd8c45d17011a2eb33532c6)

File: DKGRituals/src/mappingCoordinator.ts (hunk 2 above)
Comment (evandrosaturnino) [reply DKGRituals/src/mappingCoordinator.ts]: Addressed in [update posted transcripts and aggregations properties name](https://github.com/nucypher/taco-subgraphs/pull/10/commits/b65c94eb6e520c11d6bfea91b72c81ccb3708ca2)

**Review by manumonti — CHANGES_REQUESTED** [review]
Great work! I would suggest some changes though

File: DKGRituals/src/coordinator.ts [hunk 6]
```diff
@@ -1,131 +0,0 @@
-import { Bytes, log } from "@graphprotocol/graph-ts";
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply DKGRituals/src/coordinator.ts]: Addressed in https://github.com/nucypher/taco-subgraphs/pull/10/commits/6134a39c53509de7d9d0f9b2ffb3e8a3f460f58c

File: DKGRituals/src/utils/constants.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { BigInt, BigDecimal, Address } from "@graphprotocol/graph-ts";
+
+export const RITUAL_COUNTER_ID = "Counter";
+export const ADDRESS_ZERO = Address.fromString(
+  "0x0000000000000000000000000000000000000000",
+);
+export const COORDINATOR_CONTRACT = Address.fromString(
+  "0xE74259e3dafe30bAA8700238e324b47aC98FE755",
+);
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply DKGRituals/src/utils/constants.ts]: Addressed in [refactor: remove unused constants](https://github.com/nucypher/taco-subgraphs/pull/10/commits/1d5f731c1d348233543a6c6f054a3d21990c191c)

File: DKGRituals/src/utils/constants.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+export const MAX_REDEMPTIONS_COUNT = 1000;
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply DKGRituals/src/utils/constants.ts]: Addressed in [refactor: remove unused constants](https://github.com/nucypher/taco-subgraphs/pull/10/commits/1d5f731c1d348233543a6c6f054a3d21990c191c)

File: DKGRituals/src/utils/constants.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+export const ZERO_BI = BigInt.fromI32(0);
+export const ONE_BI = BigInt.fromI32(1);
+export const ZERO_BD = BigDecimal.fromString("0");
+export const ONE_BD = BigDecimal.fromString("1");
+export const BI_18 = BigInt.fromI32(18);
+export const SATOSHI_MULTIPLIER = BigInt.fromI64(10000000000);
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply DKGRituals/src/utils/constants.ts]: Addressed in [refactor: remove unused constants](https://github.com/nucypher/taco-subgraphs/pull/10/commits/1d5f731c1d348233543a6c6f054a3d21990c191c)

File: DKGRituals/schema.graphql [hunk 10]
```diff
@@ -29,3 +32,18 @@ type RitualCounter @entity {
   unsuccessful: Int!
   notEnded: Int!
 }
+
+type Transaction @entity {
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply DKGRituals/schema.graphql]: Addressed in [refactor: make transaction entity immutable](https://github.com/nucypher/taco-subgraphs/pull/10/commits/d928ab763721abd07277f35ba89c768376199772)

File: DKGRituals/src/mappingCoordinator.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+  transactions.push(transaction.id);
+  ritual.transactions = transactions;
+  ritual.save();
+  ritualCounter.save();
+}
+
+export function handleRitualAuthorityTransferred(
+  event: RitualAuthorityTransferredEvent,
+): void {
+  const ritual = getOrCreateRitual(event.params.ritualId.toString());
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply DKGRituals/src/mappingCoordinator.ts]: Addressed in [refactor: update coordinator mapping functions](https://github.com/nucypher/taco-subgraphs/pull/10/commits/51c2df3ab5711da43e0c6ea5449c00cb1d4b473a)

File: DKGRituals/src/mappingCoordinator.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+  postedTranscripts.push(event.params.node);
+  ritual.postedTranscripts = postedTranscripts;
+  ritual.totalPostedTranscripts = ritual.totalPostedTranscripts + 1;
+  ritual.save();
+}
+
+export function handleStartAggregationRound(
+  event: StartAggregationRoundEvent,
+): void {
+  const transaction = getOrCreateTransaction(getTransactionIDFromEvent(event));
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply DKGRituals/src/mappingCoordinator.ts]: Addressed in [refactor: update coordinator mapping functions](https://github.com/nucypher/taco-subgraphs/pull/10/commits/51c2df3ab5711da43e0c6ea5449c00cb1d4b473a)

File: DKGRituals/src/utils/helper.ts [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+import * as constants from "./constants";
+
+export function getOrCreateTransaction(id: string): Transaction {
+  let transaction = Transaction.load(id);
+  if (transaction == null) {
+    transaction = new Transaction(id);
+    transaction.id = id;
+  }
+  return transaction as Transaction;
+}
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply DKGRituals/src/utils/helper.ts]: Addressed in [refactor: update coordinator mapping functions](https://github.com/nucypher/taco-subgraphs/pull/10/commits/51c2df3ab5711da43e0c6ea5449c00cb1d4b473a)

File: DKGRituals/src/mappingCoordinator.ts [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+  const initiator = getOrCreateInitiator(ritual.initiator);
+  const rituals = initiator.rituals;
+  rituals.push(ritual.id);
+  initiator.rituals = rituals;
+  initiator.save();
+}
+
+export function handleTranscriptPosted(event: TranscriptPostedEvent): void {
+  const ritual = getOrCreateRitual(event.params.ritualId.toString());
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply DKGRituals/src/mappingCoordinator.ts]: Addressed in [refactor: update coordinator mapping functions](https://github.com/nucypher/taco-subgraphs/pull/10/commits/51c2df3ab5711da43e0c6ea5449c00cb1d4b473a)

File: DKGRituals/src/mappingCoordinator.ts [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+  ritual.authority = event.params.authority;
+  ritual.dkgSize = ritualInst.getDkgSize();
+  ritual.threshold = ritualInst.getThreshold();
+  ritual.accessController = ritualInst.getAccessController();
+  ritual.participants = changetype<Bytes[]>(event.params.participants);
+  ritual.dkgStatus = "DKG_AWAITING_TRANSCRIPTS";
+
+  const transactions = ritual.transactions;
+  transactions.push(transaction.id);
+  ritual.transactions = transactions;
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply DKGRituals/src/mappingCoordinator.ts]: Addressed in [refactor: update coordinator mapping functions](https://github.com/nucypher/taco-subgraphs/pull/10/commits/51c2df3ab5711da43e0c6ea5449c00cb1d4b473a)

File: DKGRituals/src/mappingCoordinator.ts [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+
+  const transaction = getOrCreateTransaction(getTransactionIDFromEvent(event));
+  transaction.txHash = event.transaction.hash;
+  transaction.timestamp = event.block.timestamp;
+  transaction.from = event.transaction.from;
+  transaction.to = event.transaction.to;
+  transaction.description = "Start Ritual";
+  transaction.save();
+
+  const ritual = getOrCreateRitual(event.params.ritualId.toString());
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply DKGRituals/src/mappingCoordinator.ts]: Addressed in [refactor: update coordinator mapping functions](https://github.com/nucypher/taco-subgraphs/pull/10/commits/51c2df3ab5711da43e0c6ea5449c00cb1d4b473a)

File: DKGRituals/src/utils/helper.ts [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+    ritual.participants = [];
+    ritual.totalPostedTranscripts = 0;
+    ritual.postedTranscripts = [];
+    ritual.totalPostedAggregations = 0;
+    ritual.postedAggregations = [];
+    ritual.transactions = [];
+    ritual.dkgStatus = "UNKNOWN";
+  }
+  return ritual;
+}
```
[the remark this answers is not in the document]
Comment (evandrosaturnino) [reply DKGRituals/src/utils/helper.ts]: Addressed in [refactor: update coordinator mapping functions](https://github.com/nucypher/taco-subgraphs/pull/10/commits/51c2df3ab5711da43e0c6ea5449c00cb1d4b473a)

**Review by manumonti — APPROVED** [review]
LGTM! Thanks for addressing these

**Comment by manumonti** [comment]
@evandrosaturnino 

This is the URL of the deployment of this new version (0.1.0):

https://subgraph.satsuma-prod.com/735cd3ac7b23/nucypher-ops/DKGRituals/version/0.1.0/api

I'll be checking if everything works fine. After that, I'll promote this release to Live, so the URL will be

https://subgraph.satsuma-prod.com/735cd3ac7b23/nucypher-ops/DKGRituals/api

**Pull request merged**